.github/scripts/analyze-builds.js

```diff
@@ -0,0 +1,112 @@
+    buildDirectory: "dist",
+    bundleDirectory: "js"
+  },
+  "rollup": {
+    name: "Rollup",
+    package: "rollup",
+    devDep: true,
+    buildDirectory: "public",
+    bundleDirectory: "./"
+  },
+  "webpack": {
+    name: "Webpack",
+    package: "webpack",
+    devDep: true,
+    buildDirectory: "dist",
+    bundleDirectory: "./"
+  }
+};
+
+const getDirectories = async (directoriesPath) =>
+  (await readdir(directoriesPath, { withFileTypes: true }))
+    .filter((dirent) => dirent.isDirectory() && dirent.name.charAt(0) !== ".")
+    .map((dirent) => dirent.name);
+
+(async () => {
+  try {
+    const sampleDirs = await getDirectories(SAMPLES_PATH);
+
+    const jsapiVersion = JSON.parse(
+      await readFile(resolve(__dirname, SAMPLES_PATH, sampleDirs[0], "package.json"), "utf8")
+    ).dependencies["@arcgis/core"].replace(/\^|\~/, "");
+
+    console.log(`JSAPI version: ${jsapiVersion}`);
+    const outputPath = resolve(__dirname, "../build-sizes", `${jsapiVersion}.csv`);
+    const stream = createWriteStream(outputPath);
+    stream.write("Sample,Main bundle size,On-disk size\n");
+
+    for (sample of sampleDirs) {
+      const buildDir = SAMPLES_INFO[sample]?.buildDirectory;
+      const bundleDir = SAMPLES_INFO[sample]?.bundleDirectory;
+      const sampleName = SAMPLES_INFO[sample]?.name;
+      const samplePackage = SAMPLES_INFO[sample]?.package;
+      const isPackageDevDep = SAMPLES_INFO[sample]?.devDep;
+
+      if (!!buildDir) {
+        const samplePath = resolve(SAMPLES_PATH, sample);
+        const buildPath = resolve(samplePath, buildDir);
+
+        const samplePackageFile = JSON.parse(await readFile(resolve(samplePath, "package.json"), "utf8"));
+
+        const packageVersion = (
+          !!isPackageDevDep
+            ? samplePackageFile.devDependencies[samplePackage]
+            : samplePackageFile.dependencies[samplePackage]
+        ).replace(/\^|\~/, "");
+
+        console.log(`${sampleName}: installing deps`);
+        await exec(`npm i --prefix ${samplePath}`);
+
+        console.log(`${sampleName}: building`);
+        await exec(`npm run build --prefix ${samplePath}`);
+
+        console.log(`${sampleName}: calculating size`);
+        const buildSize = (await exec(`du -sh ${buildPath} | cut -f1`)).stdout.trim();
+        const fileCount = (await exec(`find ${buildPath} -type f | wc -l`)).stdout.trim();
+        const mainBundleSize = Number(
+          (
+            await exec(
+              `find ${resolve(
+                buildPath,
+                bundleDir
+              )} -name '*.js' -type f -printf '%s\t%p\n' | sort -nr | head -1 | cut -f1`
+            )
+          ).stdout.trim() /  1e+6 // convert bytes to megabytes
+        )
+          .toFixed(1)
+          .toString()
+          .concat("M");
+
+        stream.write(`${sampleName} ${packageVersion},${mainBundleSize},${buildSize} (${fileCount} files)\n`);
+      }
+    }
+  } catch (err) {
+    console.error(err);

```

Comment:
Is there more we can do if there's an error?

Reply:
Sure, any ideas? As of now the error will be logged and the action will fail. People can set up their [notifications settings](https://github.com/settings/notifications) to receive emails when actions fail, and then they could check the action's logs.
![image](https://user-images.githubusercontent.com/10986395/152585988-630bd216-2ab2-451c-a782-88019e53f528.png)

Alternatively I could log the error to a file, or something else?

Reply:
> As of now the error will be logged and the action will fail.

Oh, that sounds good then 😄 I thought this wasn't going to anywhere that'd be noticed.

Reply:
@benelan will we get the full console log dump, or is it abstracted in some way?

Reply:
Well [this is what happens](https://github.com/benelan/test/runs/5071475460?check_suite_focus=true#step:5:5) when I did `throw new Error("test")`. Do you want me to test failing in any other ways? (Also I was wrong, I had to manually set the exit code)